src/Func.cpp

```diff
@@ -863,6 +865,108 @@ void Func::invalidate_cache() {
     }
 }
 
+Func Func::in(const Func &f) {
+    user_assert(name() != f.name()) << "Cannot call 'in()' on itself\n";
+    const map<string, IntrusivePtr<FunctionContents>> &wrappers = func.wrappers();
+    const auto &iter = wrappers.find(f.name());
+    if (iter == wrappers.end()) {
+        Func wrapper(name() + "_in_" + f.name());
+        wrapper(args()) = (*this)(args());
+        wrapper.func.freeze();
+        func.add_wrapper(wrapper.func, f.name());
+        return wrapper;
+    }
+
+    IntrusivePtr<FunctionContents> wfunc = iter->second;
+    internal_assert(wfunc.defined());
+
+    // Make sure that no other Func shares the same wrapper as 'f'
+    for (const auto &it : wrappers) {
+        if (it.first == f.name()) {
+            continue;
+        }
+        user_assert(!it.second.same_as(wfunc))
+            << "Redefinition of shared wrapper with " << it.first << " [" << name() << " -> "
+            << Function(wfunc).name() << "] in " << f.name() << " is not allowed\n";
+    }
+    Function wrapper(wfunc);
+    internal_assert(wrapper.frozen());
+    return Func(wrapper);
+}
+
+Func Func::in(const vector<Func>& fs) {
+    if (fs.empty()) {
+        return in(); // Make a global wrapper
```

Comment:
Are we sure this is the behavior we want? (ie, should it be an error condition to pass in an empty vector?) Not clear to me that the fallback should be to a global wrapper.